Add docs Array.groupByToMap (#13171)
mdn/content: files/en-us/web/javascript/reference/global_objects/array/groupbytomap/index.md

groupByToMap(callbackFn)
groupByToMap(callbackFn, thisArg)

// Inline callback function
groupByToMap(function(element) { /* ... */ })
groupByToMap(function(element, index) { /* ... */ })
groupByToMap(function(element, index, array){ /* ... */ })
groupByToMap(function(element, index, array) { /* ... */ }, thisArg)
```

### Parameters

- `callbackFn`

  - : Function to execute on each element in the array, taking 3 arguments:

    - `element`
      - : The value of the current element in the array.
    - `index` {{optional_inline}}
      - : The index (position) of the current element in the array.
    - `array` {{optional_inline}}
      - : The array that `groupBy()` was called on.

    The object returned from the callback indicates the group of the current element.

- `thisArg` {{optional_inline}}
  - : Object to use as {{jsxref("Operators/this", "this")}} inside `callbackFn`.
      If not specified, `undefined` will be used.

### Return value

A {{jsxref("Map")}} object with keys for each group, each assigned to an array containing the elements of the associated group.

### Exceptions

- `TypeError`
  - : The specified callback function is not callable.

## Description

> The `groupBytoMap()` method executes the `callbackFn` function once for each index of the array, returning an object indicating the group of the associated element.

Reviewer: `groupBytoMap()` → `groupByToMap()`.

---
title: Array.prototype.groupByToMap()
slug: Web/JavaScript/Reference/Global_Objects/Array/groupByToMap
tags:
  - Array
  - groupByToMap
  - JavaScript
  - Method
  - Prototype
  - Reference
  - Experimental
browser-compat: javascript.builtins.Array.groupByToMap
---
{{JSRef}} {{SeeCompatTable}}

The **`groupByToMap()`** method groups the elements of the calling array using the values returned by a provided testing function.
The final returned {{jsxref("Map")}} uses the unique values from the test function as keys, which can be used to get the array of elements in each group.

<!-- {{EmbedInteractiveExample("pages/js/array-groupbytomap.html")}} -->

The elements in the {{jsxref("Map")}} and the original array are the same (not {{glossary("deep copy","deep copies")}}).
Changing the internal structure of the elements will be reflected in both the original array and the returned {{jsxref("Map")}}.

> An object that needs to access the groups must keep a reference to the original key; the properties of a key can be modified, but you can't use another object that just happens to have the same name and properties.

Reviewer 2: I didn't understand this at all until I read the example further down the page. I expect it's basically impossible to summarise this, so it might be better not to say it until you have space to explain it properly with an example.

Reviewer: Is it even necessary to explain any of these object reference behaviors here? This essentially explains how a `Map` works with objects as keys, so I’d rather expect to find this description on the docs page for `Map` instead.

> This method is useful when mapping related information to a specific object that might change over time.

Reviewer 2: Is it worth adding something like "...because you can modify properties of the key object and still use it as a key to retrieve its group." (assuming I've understood the point of this)?

I guess this is related also to why you might use this instead of `groupBy()`? It would be really nice to have something in this page explicitly addressing that, just to be really clear.

Author: Yes, that's pretty much the intent, as I understand it from https://bugzilla.mozilla.org/show_bug.cgi?id=1739648#c16

I think that we do need a clear statement of why you use this, and I agree with @SebastianSimon probably most of the depth of that should be in the Map docs. I'll remove/restructure as appropriate.

> If the mapping object is invariant then you could instead use {{jsxref("Array.prototype.groupBy()")}} to group the elements, representing the object with a string.

Reviewer 2: I think this or a version of it should be in the H2#Description section, not here.

Method pages should start with a short paragraph, then interactive example, then H2#Syntax, then any more prose in H2#Description. Like, say, [`filter()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) or [`concat()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat) or [`copyWithin()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/copyWithin).

## Syntax

```js
// Arrow function
groupByToMap((element) => { /* ... */ } )
groupByToMap((element, index) => { /* ... */ } )
groupByToMap((element, index, array) => { /* ... */ } )

// Callback function
groupByToMap(callbackFn)
groupByToMap(callbackFn, thisArg)

// Inline callback function
groupByToMap(function(element) { /* ... */ })
groupByToMap(function(element, index) { /* ... */ })
groupByToMap(function(element, index, array){ /* ... */ })
groupByToMap(function(element, index, array) { /* ... */ }, thisArg)
```

### Parameters

- `callbackFn`

  - : Function to execute on each element in the array, taking 3 arguments:

    - `element`
      - : The current element in the array.
    - `index` {{optional_inline}}
      - : The index (position) of the current element in the array.

>     - `array` {{optional_inline}}

Reviewer 2: In both these cases I don't think `Optional` is appropriate here. `Optional` is for arguments you don't have to _pass_, but here the arguments are always passed, and we are using it to suggest that the function doesn't have to _use_ them. But no function ever has to use any of its arguments. - we might as well say `element` is optional in this sense.

      - : The array that `groupBy()` was called on.

    The value ({{Glossary("object")}} or {{Glossary("primitive")}}) returned from the callback indicates the group of the current element.

- `thisArg` {{optional_inline}}
  - : Object to use as {{jsxref("Operators/this", "this")}} inside `callbackFn`.
      If not specified, `undefined` will be used.

### Return value

A {{jsxref("Map")}} object with keys for each group, each assigned to an array containing the elements of the associated group.

### Exceptions

- `TypeError`
  - : The specified callback function is not callable.

## Description

> The `groupByToMap()` method executes the `callbackFn` function once for each index of the array, returning a value indicating the group of the associated element.

Reviewer 2: This implies to me that "The `groupByToMap()` method" is the thing that returns "a value indicating the group of the associated element". Perhaps:

```suggestion
The `groupByToMap()` method executes the callback function once for each index of the array. The callback function returns a value indicating the group of the associated element.
```

? (also I think maybe we should consistently use either "the callback function" or "`callbackFn`" in this section)

Each key has an associated array containing all the elements for which the callback returned the same value.

`callbackFn` is invoked for _every_ index of the array, not just those with assigned values.
This means it may be less efficient for sparse arrays, compared to methods that only visit assigned values.

If a `thisArg` parameter is provided to `groupByToMap()`, it will be used as the `this` value inside each invocation of the `callbackFn`.
If it is not provided, then {{jsxref("undefined")}} is used.

The `groupByToMap()` method does not mutate the array on which it is called, but the function provided to `callbackFn` can.
Note however that the elements processed by `groupByToMap()` are set _before_ the first invocation of `callbackFn`.
Therefore:

- `callbackFn` will not visit any elements added to the array after the call to `groupByToMap()` begins.
- Elements that are assigned to indexes already visited, or to indexes outside the range, will not be visited by `callbackFn`.
- If an existing, yet-unvisited element of the array is changed by `callbackFn`, its value passed to the `callbackFn` will be the value at the time `groupByToMap()` visits that element's index.

> - Elements that are {{jsxref("Operators/delete", "deleted")}} are still visited.

Reviewer 2: ```suggestion
- Elements that are {{jsxref("Operators/delete", "deleted", "nocode")}} are still visited.
```

...since "deleted" isn't a keyword it should not get syntax highlighting. But also, is https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/delete correct here? That's not how you delete an element from an array, is it?

Author: Thanks! Yes, that is how you delete elements from an array, leaving an empty index for the element you deleted. 
Since of course you might more naturally use a method that returns a new array with specific elements remove (e.g. `splice()` I've updated the link to make this more clear.

> > **Warning:** Concurrent modifications of the kind described above frequently lead to hard-to-understand code and are generally to be avoided (except in special cases).

Reviewer 2: You might consider using a subheading for this (especially if you migrate some stuff into H2#Description from the earlier intro prose). Like "Mutating the array in the callback" or something. Partly because this is kind of edge-casey, so putting it in a separate section makes it easier for people to ignore.

Reviewer 2: Not sure if this got missed or you decided not to (I'm fine if you decided not to).

Author: Hi @wbamberg,
Sorry, I did miss this. It is a very good suggestion. Fixed in both docs.

## Examples

First we define an array containing objects representing an inventory of different foodstuffs.

> Each food has a `type`, which might need to be stored differently, and a `quantity` that can be used to determine when we need to reorder each item.

Reviewer 2: I don't understand what "which might need to be stored differently" means here. It suggests data storage perhaps when I guess you mean food storage? I think it would be OK to delete this.

Reviewer: It’s probably “storing the foodstuffs differently, as a real world application”, i.e. meats should be stored separately in a freezer, etc., but the sentence is indeed confusing, especially because it’s unclear which meaning the word “stored” has (_computer memory_ or _kitchen pantry_?); perhaps “reorder” has the same problem (_sorting_ or _restocking_?). Maybe this:

> First we define an array containing objects representing an inventory of different foodstuffs. Each food item has a `type` and a `quantity`. The `type` may indicate how a given food item might need to be stored in a kitchen pantry, and the `quantity` may be used to determine when we need to restock each item.

Maybe someone can come up with a better example usage for “store”. Maybe an explicit example helps, e.g. _“For example, `{ name: "fish", type: "meat", quantity: 22 }` means that our kitchen has 22 fish items in stock, and, since it’s meat, they need to be stored in the freezer.”_ or something like that. Maybe we can come up with a completely different example that doesn’t involve food items — and how to store them — at all.

The `groupBy` article has exactly the same example and should be edited as well.

Author: Thanks. Yes this is a problem. I will fix it. Probably using text like Sebastians.

Reviewer 2: > It’s probably “storing the foodstuffs differently, as a real world application”

That was my guess, but I didn't want to assume :). I agree that removing the line is better than trying to explain it more.

```js
const inventory = [
  { name: 'asparagus', type: 'vegetables', quantity: 9 },
  { name: 'bananas',  type: 'fruit', quantity: 5 },
  { name: 'goat', type: 'meat', quantity: 23 },
  { name: 'cherries', type: 'fruit', quantity: 12 },
  { name: 'fish', type: 'meat', quantity: 22 }
];
```

The code below uses `groupByToMap()` with an arrow function that returns the object keys named `reorder` or `sufficient`, depending on whether the element has `quantity < 6`.
The returned `result` object is a `Map` so we need to call `get()` with the key to obtain the array.

```js

> let reorder  = { reorder: true };

Reviewer 2: ```suggestion
const reorder  = { reorder: true };
```

I'm Mr Let-to-const. Although you do `reorder['fast']  = true ;`, that's allowed because `reorder` is not reassigned.

Reviewer 2: Also it's a bit confusing that the object variable is `reorder`, but it contains a boolean also called `reorder`. I don't have a great idea about how to improve this though.

Reviewer: Well, honestly, the example is strange anyway. Why would you need objects as keys here? What is the motivation behind needing to keep track of references for the keys? In which scenario is mutating the key a good idea? I don’t currently have a better alternative, either way. Just a small note: if we reword “reorder” to “restock”, this identifier needs to be updated here as well.

Author: Thanks. I made the change to restock here and everywhere. 
Yes, the example is bogus, but I suspect there is no "simple" real world simple example where you would do this. If you look at the developer explanation in https://bugzilla.mozilla.org/show_bug.cgi?id=1739648#c16 it seems clear to me that it certainly isn't something you would need to do every day :-)

If you look at line 81 in updated doc I hope the explanation of why at least you might use this makes more sense, even though no intuitive example is provided.

console.log(result.get(reorder));
// expected output: Array [Object { name: "bananas", type: "fruit", quantity: 5 }]
```

Note that the function argument `{ quantity }` is a basic example of [object destructuring syntax for function arguments](/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#unpacking_fields_from_objects_passed_as_a_function_parameter).
This unpacks the `quantity` property of an object passed as a parameter, and assigns it to a variable named `quantity` in the body of the function.
This is a very succinct way to access the relevant values of elements within a function.

An object key can be modified and still used.
However you can't recreate the key and still use it.
For this reason it is important that anything that needs to use the map keeps a reference to its keys.

```js
// A modified key can still be used in a Map
reorder['fast']  = true ;
console.log(result.get(reorder));
// expected output: Array [Object { name: "bananas", type: "fruit", quantity: 5 }]

// But you can't use a different key, even if it has the same structure!
const reorder2  = { reorder: true };
console.log(result.get(reorder2));
// expected output: undefined

> ```

Reviewer 2: Yeah, this is a very nice explanation of the confusing paragraph at the start.

> The callback syntax provides access to the array and current index, so it is possible to implement grouping strategies based on the values of other elements in the array.

Reviewer 2: This is interesting, and might be better in H2#Description than in H2#Examples.

Author: I had it here to indicate "this is a simple example - you can do a bit more if you like". As you say, probably better in the description. Text added

> `callbackFn` is called with the value of the current element, the current index, and the array itself.
While groups often depend only on the current element, it is possible to implement grouping strategies based on the values of other elements in the array.

## Specifications

{{Specifications}}

## Browser compatibility

{{Compat}}

## See also

- {{jsxref("Array.prototype.groupBy()")}}

> - [Polyfill of `Array.prototype.groupBy` in `core-js`](https://github.com/zloirock/core-js#array-grouping)

Reviewer 2: ```suggestion
- [Polyfill of `Array.prototype.groupByToMap()` in `core-js`](https://github.com/zloirock/core-js#array-grouping)
```
?

Reviewer: Are the parentheses actually needed?

Reviewer 3: Yes, we put parenthesis after method names to distinguish the from properties/attributes.

Reviewer 2: It seems like this change is not applied?

---
title: Array.prototype.groupByToMap()
slug: Web/JavaScript/Reference/Global_Objects/Array/groupByToMap
tags:
  - Array
  - groupByToMap
  - JavaScript
  - Method
  - Prototype
  - Reference
  - Experimental
browser-compat: javascript.builtins.Array.groupByToMap
---
{{JSRef}} {{SeeCompatTable}}

The **`groupByToMap()`** method groups the elements of the calling array using the values returned by a provided testing function.
The final returned {{jsxref("Map")}} uses the unique values from the test function as keys, which can be used to get the array of elements in each group.

<!-- {{EmbedInteractiveExample("pages/js/array-groupbytomap.html")}} -->

> The method is primarily useful when grouping elements that are associated with an object, and in particular when that object might change over time.

Author: @wbamberg I've moved most of the info out of the overview down to the description - here just a summary and alternative.

If the object is invariant, you might instead represent it using a string, and group elements with {{jsxref("Array.prototype.groupBy()")}}.

## Syntax

```js
// Arrow function
groupByToMap((element) => { /* ... */ } )
groupByToMap((element, index) => { /* ... */ } )
groupByToMap((element, index, array) => { /* ... */ } )

// Callback function
groupByToMap(callbackFn)
groupByToMap(callbackFn, thisArg)

// Inline callback function
groupByToMap(function(element) { /* ... */ })
groupByToMap(function(element, index) { /* ... */ })
groupByToMap(function(element, index, array){ /* ... */ })
groupByToMap(function(element, index, array) { /* ... */ }, thisArg)
```

### Parameters

- `callbackFn`

  - : Function to execute on each element in the array, taking 3 arguments:

    - `element`
      - : The current element in the array.
    - `index`
      - : The index (position) of the current element in the array.
    - `array`

>       - : The array that `groupBy()` was called on.

Reviewer 2: ```suggestion
      - : The array that `groupByToMap()` was called on.
```

## Description

The `groupByToMap()` method executes `callbackFn` once for each index of the array.
The callback function returns a value indicating the group of the associated element.
The values returned by `callbackFn` are used as keys for the {{jsxref("Map")}} returned by `groupByToMap()`.
Each key has an associated array containing all the elements for which the callback returned the same value.

The elements in the returned {{jsxref("Map")}} and the original array are the same (not {{glossary("deep copy","deep copies")}}).
Changing the internal structure of the elements will be reflected in both the original array and the returned {{jsxref("Map")}}.

This method is useful when you need to group information that is related to a particular object that might potentially change over time.
This is because even if the object is modified, it will continue to work as a key to the returned `Map`.
If you instead create a string representation for the object and use that as a grouping key in {{jsxref("Array.prototype.groupBy()")}}, you must maintain the mapping between the original object and its representation as the object changes.

> > **Note:** An object that needs to access the groups must keep a reference to the _original_ key (although you may modify its properties).

Reviewer 2: I think "An object" here is confusing, since we are talking about the objects used as keys already, but this is just "the caller". Maybe something like: 

```suggestion
> **Note:** To access the groups in the returned `Map`, you must use the same object that was originally used as a key in the `Map` (although you may modify its properties).
```

Author: Yes. Thanks!

> You can't use another object that just happens to have the same name and properties.

`callbackFn` is called with the value of the current element, the current index, and the array itself.
While groups often depend only on the current element, it is possible to implement grouping strategies based on the values of other elements in the array.

`callbackFn` is invoked for _every_ index of the array, not just those with assigned values.
This means it may be less efficient for sparse arrays, compared to methods that only visit assigned values.

If a `thisArg` parameter is provided to `groupByToMap()`, it will be used as the `this` value inside each invocation of the `callbackFn`.

> If it is not provided, then {{jsxref("undefined")}} is used.

Reviewer 2: This does not appear to be correct. Trying our the example, with extra logging:

```js
const inventory = [
  { name: 'asparagus', type: 'vegetables', quantity: 9 },
  { name: 'bananas',  type: 'fruit', quantity: 5 },
  { name: 'goat', type: 'meat', quantity: 23 },
  { name: 'cherries', type: 'fruit', quantity: 12 },
  { name: 'fish', type: 'meat', quantity: 22 }
];

let restock  = { restock: true };
const sufficient = { restock: false };
const result = inventory.groupByToMap( ({ quantity }) => {
  console.log(`it is: ${this}`)
  return quantity < 6 ? restock : sufficient
});
console.log(result.get(restock));
```

...I get:

```
it is: [object Window]
```
```

Author: @wbamberg Interesting! This is what the [spec says](https://tc39.es/proposal-array-grouping/#sec-array.prototype.groupbymap), so I propose that we merge this.

I will of course still chase this in https://bugzilla.mozilla.org/show_bug.cgi?id=1739648#c17

Reviewer 2: Yes, I agree we should document the spec, and thanks for pursuing this in the bug!

Author: Finally https://github.com/mdn/content/pull/14421

- Elements that are [deleted](/en-US/docs/Web/JavaScript/Reference/Operators/delete#deleting_array_elements) are still visited.

> **Warning:** Concurrent modifications of the kind described above frequently lead to hard-to-understand code and are generally to be avoided (except in special cases).

## Examples

First we define an array containing objects representing an inventory of different foodstuffs.
Each food has a `type` and a `quantity`.

```js
const inventory = [
  { name: 'asparagus', type: 'vegetables', quantity: 9 },
  { name: 'bananas',  type: 'fruit', quantity: 5 },
  { name: 'goat', type: 'meat', quantity: 23 },
  { name: 'cherries', type: 'fruit', quantity: 12 },
  { name: 'fish', type: 'meat', quantity: 22 }
];
```

The code below uses `groupByToMap()` with an arrow function that returns the object keys named `restock` or `sufficient`, depending on whether the element has `quantity < 6`.
The returned `result` object is a `Map` so we need to call `get()` with the key to obtain the array.

```js

> let restock  = { restock: true };

Reviewer 2: This is still `let` and should be `const`, per https://github.com/mdn/content/pull/13171#discussion_r814467008.
